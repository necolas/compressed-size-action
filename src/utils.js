import fs from 'fs';

function prettyBytes(bytes, toFixed) {
  var kb = bytes / 1000;
  return kb.toFixed(toFixed || 2);
}

/**
 * Check if a given file exists and can be accessed.
 * @param {string} filename
 */
export async function fileExists(filename) {
	try {
		await fs.promises.access(filename, fs.constants.F_OK);
		return true;
	} catch (e) {}
	return false;
}

/**
 * Remove any matched hash patterns from a filename string.
 * @param {string=} regex
 * @returns {(((fileName: string) => string) | undefined)}
 */
export function stripHash(regex) {
	if (regex) {
		console.log(`Striping hash from build chunks using '${regex}' pattern.`);
		return function (fileName) {
			return fileName.replace(new RegExp(regex), (str, ...hashes) => {
				hashes = hashes.slice(0, -2).filter((c) => c != null);
				if (hashes.length) {
					for (let i = 0; i < hashes.length; i++) {
						const hash = hashes[i] || '';
						str = str.replace(hash, hash.replace(/./g, '*'));
					}
					return str;
				}
				return '';
			});
		};
	}

	return undefined;
}

/**
 * @param {number} delta
 * @param {number} difference
 */
export function getDeltaText(delta) {
	const deltaText = (delta > 0 ? '+' : '') + prettyBytes(delta);
	return deltaText;
}

function getDifference(before, delta) {
	const divisor = before === 0 ? Math.abs(delta) : before
	return ((delta / divisor) * 100).toFixed(1);
}

function getDifferenceText(difference) {
	const text = (difference > 0 ? '+' : '') + difference + '%';
	return text;
}

/**
 * @param {number} difference
 */
export function iconForDifference(difference) {
	let icon = '';
	if (difference >= 20) icon = 'ðŸ”´';
	else if (difference >= 10) icon = 'ðŸŸ¡';
	else if (difference <= -5) icon = 'ðŸŸ¢';
	return icon;
}

/**
 * Create a Markdown table from text rows
 * @param {string[]} rows
 */
function markdownTable(rows) {
	if (rows.length == 0) {
		return '';
	}

	// Skip all empty columns
	while (rows.every(columns => !columns[columns.length - 1])) {
		for (const columns of rows) {
			columns.pop();
		}
	}

	const [firstRow] = rows;
	const columnLength = firstRow.length;
	if (columnLength === 0) {
		return '';
	}

	return [
		// Header
		['Filename', 'Size (kB)', 'kB change', '% change', ''].slice(0, columnLength),
		// Align
		[':---', ':---', ':---', ':---', ':---:'].slice(0, columnLength),
		// Body
		...rows
	].map(columns => `| ${columns.join(' | ')} |`).join('\n');
}

/**
 * @typedef {Object} Diff
 * @property {string} filename
 * @property {number} size
 * @property {number} delta
 */

/**
 * Create a Markdown table showing diff data
 * @param {Diff[]} files
 * @param {object} options
 * @param {boolean} [options.showTotal]
 * @param {boolean} [options.collapseUnchanged]
 * @param {boolean} [options.omitUnchanged]
 * @param {number} [options.minimumChangeThreshold]
 */
export function diffTable(files, { showTotal, collapseUnchanged, omitUnchanged, minimumChangeThreshold }) {
	let changedRows = [];
	let unChangedRows = [];

	let totalSize = 0;
	let totalDelta = 0;
	for (const file of files) {
		const {
			filename,
			compressedSize,
			compressedSizeBefore,
			compressedSizeDelta,
			size,
			sizeBefore,
			sizeDelta
		} = file;
		totalSize += compressedSize;
		totalDelta += compressedSizeDelta;
		const compressedDifference = getDifference(compressedSizeBefore, compressedSizeDelta);
		const sizeDifference = getDifference(sizeBefore, sizeDelta);
		const isUnchanged = Math.abs(compressedSizeDelta) < minimumChangeThreshold;

		if (isUnchanged && omitUnchanged) continue;

		const columns = [
			// file name
			`\`${filename}\``,
			// size before
			// `**${prettyBytes(compressedSizeBefore)}**&nbsp;(${prettyBytes(sizeBefore)})`,
			// size
			`**${prettyBytes(compressedSize)}**&nbsp;(${prettyBytes(size)})`,
			// absolute change
			`**${getDeltaText(compressedSizeDelta)}**&nbsp;(${getDeltaText(sizeDelta)})`,
			// percentage change
			`**${getDifferenceText(compressedDifference)}**&nbsp;(${getDifferenceText(sizeDifference)})`,
			// icon
			iconForDifference(compressedSizeDelta)
		];
		if (isUnchanged && collapseUnchanged) {
			unChangedRows.push(columns);
		} else {
			changedRows.push(columns);
		}
	}

	let out = markdownTable(changedRows);

	if (unChangedRows.length !== 0) {
		const outUnchanged = markdownTable(unChangedRows);
		out += `\n\n<details><summary><strong>View unchanged</strong></summary>\n\n${outUnchanged}\n\n</details>\n\n`;
	}

	if (showTotal) {
		const totalDifference = ((totalDelta / totalSize) * 100) | 0;
		let totalDeltaText = getDeltaText(totalDelta, totalDifference);
		let totalIcon = iconForDifference(totalDifference);
		out = `**Total size:** ${prettyBytes(totalSize)} kB\n\n${out}`;
		out = `**Size change:** ${totalDeltaText} kB ${totalIcon}\n${out}`;
	}

	return out;
}

/**
 * Convert a string "true"/"yes"/"1" argument value to a boolean
 * @param {string} v
 */
export function toBool(v) {
	return /^(1|true|yes)$/.test(v);
}
