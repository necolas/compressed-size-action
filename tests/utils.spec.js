import { toBool, getDeltaText, iconForDifference, diffTable, fileExists, stripHash } from '../src/utils.js';

test('toBool', () => {
	expect(toBool('1')).toBe(true);
	expect(toBool('true')).toBe(true);
	expect(toBool('yes')).toBe(true);

	expect(toBool('0')).toBe(false);
	expect(toBool('false')).toBe(false);
	expect(toBool('no')).toBe(false);
});

test('getDeltaText', () => {
	expect(getDeltaText(5000, 25.5)).toMatchSnapshot();
	expect(getDeltaText(-5000, -25.5)).toMatchSnapshot();
	expect(getDeltaText(0, 0)).toMatchSnapshot();
});

test('iconForDifference', () => {
	expect(iconForDifference(0)).toBe('');
});

test('diffTable', () => {
	const files = [
		// increase
		{
			filename: 'increase.js',
			compressedSize: 5000,
			compressedSizeBefore: 2500,
			compressedSizeDelta: 2500,
			size: 10000,
			sizeBefore: 7000,
			sizeDelta: 3000,
		},
		// decrease
		{
			filename: 'decrease.js',
			compressedSize: 2500,
			compressedSizeBefore: 5000,
			compressedSizeDelta: -2500,
			size: 7500,
			sizeBefore: 12000,
			sizeDelta: -4500
		},
		// unchanged
		{
			filename: 'unchanged.js',
			compressedSize: 300,
			compressedSizeBefore: 300,
			compressedSizeDelta: 0,
			size: 1000,
			sizeBefore: 1000,
			sizeDelta: 0
		},
		// unchanged
		{
			filename: 'unchanged-tiny.js',
			compressedSize: 0,
			compressedSizeBefore: 0,
			compressedSizeDelta: 0,
			size: 10,
			sizeBefore: 10,
			sizeDelta: 0
		},
		// added
		{
			filename: 'added.js',
			compressedSize: 5000,
			compressedSizeBefore: 0,
			compressedSizeDelta: 5000,
			size: 8000,
			sizeBefore: 0,
			sizeDelta: 8000
		},
		// removed
		{
			filename: 'removed.js',
			compressedSize: 0,
			compressedSizeBefore: 5000,
			compressedSizeDelta: -5000,
			size: 0,
			sizeBefore: 12000,
			sizeDelta: -12000
		},
		// threshold
		{
			filename: 'threshold.js',
			compressedSize: 4500,
			compressedSizeBefore: 4491,
			compressedSizeDelta: 9,
			size: 7500,
			sizeBefore: 7491,
			sizeDelta: 9
		},
	];
	const defaultOptions = {
		showTotal: true,
		collapseUnchanged: true,
		omitUnchanged: false,
		minimumChangeThreshold: 1
	};

	expect(diffTable(files, { ...defaultOptions })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, showTotal: false })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, collapseUnchanged: false })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, omitUnchanged: true })).toMatchSnapshot();
	expect(diffTable(files, { ...defaultOptions, minimumChangeThreshold: 10 })).toMatchSnapshot();
	expect(diffTable(files.map(file => ({...file, delta: 0})), { ...defaultOptions })).toMatchSnapshot();

	expect(diffTable([files[2]], { ...defaultOptions })).toMatchSnapshot();
});

test('fileExists', async () => {
	expect(await fileExists('package.json')).toBe(true);
	expect(await fileExists('file-that-does-not-exist')).toBe(false);
});

test('stripHash', () => {
	expect(stripHash('\\b\\w{5}\\.')('foo.abcde.js')).toBe('foo.js');
	expect(stripHash('\\.(\\w{5})\\.chunk\\.js$')('foo.abcde.chunk.js')).toBe('foo.*****.chunk.js');
	expect(stripHash('')).toBe(undefined);
});
