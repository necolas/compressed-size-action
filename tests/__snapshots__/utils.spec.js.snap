// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffTable 1`] = `
"**Size change:** +0.01 kB 
**Total size:** 17.30 kB

| Filename | Size (kB) | kB change | % change |  |
| :--- | :--- | :--- | :--- | :---: |
| \`increase.js\` | **5.00**&nbsp;(10.00) | **+2.50**&nbsp;(+3.00) | **+100.0%**&nbsp;(+42.9%) | 游댮 |
| \`decrease.js\` | **2.50**&nbsp;(7.50) | **-2.50**&nbsp;(-4.50) | **-50.0%**&nbsp;(-37.5%) | 游릭 |
| \`added.js\` | **5.00**&nbsp;(8.00) | **+5.00**&nbsp;(+8.00) | **+100.0%**&nbsp;(+100.0%) | 游댮 |
| \`removed.js\` | **0.00**&nbsp;(0.00) | **-5.00**&nbsp;(-12.00) | **-100.0%**&nbsp;(-100.0%) | 游릭 |
| \`threshold.js\` | **4.50**&nbsp;(7.50) | **+0.01**&nbsp;(+0.01) | **+0.2%**&nbsp;(+0.1%) |  |

<details><summary><strong>View unchanged</strong></summary>

| Filename | Size (kB) | kB change | % change |
| :--- | :--- | :--- | :--- |
| \`unchanged.js\` | **0.30**&nbsp;(1.00) | **0.00**&nbsp;(0.00) | **0.0%**&nbsp;(0.0%) |

</details>

"
`;

exports[`diffTable 2`] = `
"| Filename | Size (kB) | kB change | % change |  |
| :--- | :--- | :--- | :--- | :---: |
| \`increase.js\` | **5.00**&nbsp;(10.00) | **+2.50**&nbsp;(+3.00) | **+100.0%**&nbsp;(+42.9%) | 游댮 |
| \`decrease.js\` | **2.50**&nbsp;(7.50) | **-2.50**&nbsp;(-4.50) | **-50.0%**&nbsp;(-37.5%) | 游릭 |
| \`added.js\` | **5.00**&nbsp;(8.00) | **+5.00**&nbsp;(+8.00) | **+100.0%**&nbsp;(+100.0%) | 游댮 |
| \`removed.js\` | **0.00**&nbsp;(0.00) | **-5.00**&nbsp;(-12.00) | **-100.0%**&nbsp;(-100.0%) | 游릭 |
| \`threshold.js\` | **4.50**&nbsp;(7.50) | **+0.01**&nbsp;(+0.01) | **+0.2%**&nbsp;(+0.1%) |  |

<details><summary><strong>View unchanged</strong></summary>

| Filename | Size (kB) | kB change | % change |
| :--- | :--- | :--- | :--- |
| \`unchanged.js\` | **0.30**&nbsp;(1.00) | **0.00**&nbsp;(0.00) | **0.0%**&nbsp;(0.0%) |

</details>

"
`;

exports[`diffTable 3`] = `
"**Size change:** +0.01 kB 
**Total size:** 17.30 kB

| Filename | Size (kB) | kB change | % change |  |
| :--- | :--- | :--- | :--- | :---: |
| \`increase.js\` | **5.00**&nbsp;(10.00) | **+2.50**&nbsp;(+3.00) | **+100.0%**&nbsp;(+42.9%) | 游댮 |
| \`decrease.js\` | **2.50**&nbsp;(7.50) | **-2.50**&nbsp;(-4.50) | **-50.0%**&nbsp;(-37.5%) | 游릭 |
| \`unchanged.js\` | **0.30**&nbsp;(1.00) | **0.00**&nbsp;(0.00) | **0.0%**&nbsp;(0.0%) |  |
| \`added.js\` | **5.00**&nbsp;(8.00) | **+5.00**&nbsp;(+8.00) | **+100.0%**&nbsp;(+100.0%) | 游댮 |
| \`removed.js\` | **0.00**&nbsp;(0.00) | **-5.00**&nbsp;(-12.00) | **-100.0%**&nbsp;(-100.0%) | 游릭 |
| \`threshold.js\` | **4.50**&nbsp;(7.50) | **+0.01**&nbsp;(+0.01) | **+0.2%**&nbsp;(+0.1%) |  |"
`;

exports[`diffTable 4`] = `
"**Size change:** +0.01 kB 
**Total size:** 17.30 kB

| Filename | Size (kB) | kB change | % change |  |
| :--- | :--- | :--- | :--- | :---: |
| \`increase.js\` | **5.00**&nbsp;(10.00) | **+2.50**&nbsp;(+3.00) | **+100.0%**&nbsp;(+42.9%) | 游댮 |
| \`decrease.js\` | **2.50**&nbsp;(7.50) | **-2.50**&nbsp;(-4.50) | **-50.0%**&nbsp;(-37.5%) | 游릭 |
| \`added.js\` | **5.00**&nbsp;(8.00) | **+5.00**&nbsp;(+8.00) | **+100.0%**&nbsp;(+100.0%) | 游댮 |
| \`removed.js\` | **0.00**&nbsp;(0.00) | **-5.00**&nbsp;(-12.00) | **-100.0%**&nbsp;(-100.0%) | 游릭 |
| \`threshold.js\` | **4.50**&nbsp;(7.50) | **+0.01**&nbsp;(+0.01) | **+0.2%**&nbsp;(+0.1%) |  |"
`;

exports[`diffTable 5`] = `
"**Size change:** +0.01 kB 
**Total size:** 17.30 kB

| Filename | Size (kB) | kB change | % change |  |
| :--- | :--- | :--- | :--- | :---: |
| \`increase.js\` | **5.00**&nbsp;(10.00) | **+2.50**&nbsp;(+3.00) | **+100.0%**&nbsp;(+42.9%) | 游댮 |
| \`decrease.js\` | **2.50**&nbsp;(7.50) | **-2.50**&nbsp;(-4.50) | **-50.0%**&nbsp;(-37.5%) | 游릭 |
| \`added.js\` | **5.00**&nbsp;(8.00) | **+5.00**&nbsp;(+8.00) | **+100.0%**&nbsp;(+100.0%) | 游댮 |
| \`removed.js\` | **0.00**&nbsp;(0.00) | **-5.00**&nbsp;(-12.00) | **-100.0%**&nbsp;(-100.0%) | 游릭 |

<details><summary><strong>View unchanged</strong></summary>

| Filename | Size (kB) | kB change | % change |
| :--- | :--- | :--- | :--- |
| \`unchanged.js\` | **0.30**&nbsp;(1.00) | **0.00**&nbsp;(0.00) | **0.0%**&nbsp;(0.0%) |
| \`threshold.js\` | **4.50**&nbsp;(7.50) | **+0.01**&nbsp;(+0.01) | **+0.2%**&nbsp;(+0.1%) |

</details>

"
`;

exports[`diffTable 6`] = `
"**Size change:** +0.01 kB 
**Total size:** 17.30 kB

| Filename | Size (kB) | kB change | % change |  |
| :--- | :--- | :--- | :--- | :---: |
| \`increase.js\` | **5.00**&nbsp;(10.00) | **+2.50**&nbsp;(+3.00) | **+100.0%**&nbsp;(+42.9%) | 游댮 |
| \`decrease.js\` | **2.50**&nbsp;(7.50) | **-2.50**&nbsp;(-4.50) | **-50.0%**&nbsp;(-37.5%) | 游릭 |
| \`added.js\` | **5.00**&nbsp;(8.00) | **+5.00**&nbsp;(+8.00) | **+100.0%**&nbsp;(+100.0%) | 游댮 |
| \`removed.js\` | **0.00**&nbsp;(0.00) | **-5.00**&nbsp;(-12.00) | **-100.0%**&nbsp;(-100.0%) | 游릭 |
| \`threshold.js\` | **4.50**&nbsp;(7.50) | **+0.01**&nbsp;(+0.01) | **+0.2%**&nbsp;(+0.1%) |  |

<details><summary><strong>View unchanged</strong></summary>

| Filename | Size (kB) | kB change | % change |
| :--- | :--- | :--- | :--- |
| \`unchanged.js\` | **0.30**&nbsp;(1.00) | **0.00**&nbsp;(0.00) | **0.0%**&nbsp;(0.0%) |

</details>

"
`;

exports[`diffTable 7`] = `
"**Size change:** 0.00 kB 
**Total size:** 0.30 kB



<details><summary><strong>View unchanged</strong></summary>

| Filename | Size (kB) | kB change | % change |
| :--- | :--- | :--- | :--- |
| \`unchanged.js\` | **0.30**&nbsp;(1.00) | **0.00**&nbsp;(0.00) | **0.0%**&nbsp;(0.0%) |

</details>

"
`;

exports[`getDeltaText 1`] = `"+5.00"`;

exports[`getDeltaText 2`] = `"-5.00"`;

exports[`getDeltaText 3`] = `"0.00"`;
